cmd/server/search.go

```diff
@@ -290,6 +291,114 @@ func (s *searcher) TopAltNames(limit int, minMatch float64, alt string) []Alt {
+	}
+
+	indexedTokens := strings.Fields(indexed)
+	searchTokensLength := sumLength(searchTokens)
+	indexTokensLength := sumLength(indexedTokens)
+
+	//Compare each search token to each indexed token. Sort the results in descending order
+	scores := make([]Score, 0)
+	for searchIdx, searchToken := range searchTokens {
+		for indexIdx, indexedToken := range indexedTokens {
+			score := customJaroWinkler(indexedToken, searchToken)
+			scores = append(scores, Score{score, searchIdx, indexIdx})
+		}
+	}
+	sort.Slice(scores[:], func(i, j int) bool {
+		return scores[i].score > scores[j].score
+	})
+
+	//Pick the highest score for each search term, where the indexed token hasn't yet been matched
+	matchedSearchTokens := make([]bool, len(searchTokens))
+	matchedIndexTokens := make([]bool, len(indexedTokens))
+	matchedIndexTokensLength := 0
+	totalWeightedScores := 0.0
+	for _, score := range scores {
+		//If neither the search token nor index token have been matched so far
+		if matchedSearchTokens[score.searchTokenIdx] == false && matchedIndexTokens[score.indexTokenIdx] == false {
+			//Weight the importance of this word score by its character length
+			searchToken := searchTokens[score.searchTokenIdx]
+			indexToken := indexedTokens[score.indexTokenIdx]
+			totalWeightedScores += score.score * float64(len(searchToken)+len(indexToken))
+
+			matchedSearchTokens[score.searchTokenIdx] = true
+			matchedIndexTokens[score.indexTokenIdx] = true
+			matchedIndexTokensLength += len(indexToken)
+		}
+	}
+	lengthWeightedAverageScore := totalWeightedScores / float64(searchTokensLength+matchedIndexTokensLength)
+
+	//If some index tokens weren't matched by any search token, penalise this search a small amount. If this isn't done,
+	//a query of "John Doe" will match "John Doe" and "John Bartholomew Doe" equally well.
+	//Calculate the fraction of the index name that wasn't matched, apply a weighting to reduce the importance of
+	//unmatched portion, then scale down the final score.
+	matchedIndexLength := 0
+	for i, str := range indexedTokens {
+		if matchedIndexTokens[i] == true {
+			matchedIndexLength += len(str)
+		}
+	}
+	matchedFraction := float64(matchedIndexLength) / float64(indexTokensLength)
+	return lengthWeightedAverageScore * scalingFactor(matchedFraction, unmatchedIndexPenaltyWeight)
+}
+
+func customJaroWinkler(s1 string, s2 string) float64 {
+	score := smetrics.JaroWinkler(s1, s2, boostThreshold, prefixSize)
+
+	if lengthMetric := lengthDifferenceFactor(s1, s2); lengthMetric < lengthDifferenceCutoffFactor {
+		//If there's a big difference in matched token lengths, punish the score. Jaro-Winkler is quite permissive about
+		//different lengths
+		score = score * scalingFactor(lengthMetric, lengthDifferencePenaltyWeight)
+	}
+	if s1[0] != s2[0] {
+		//Penalise words that start with a different characters. Jaro-Winkler is too lenient on this
+		//TODO should use a phonetic comparison here, like Soundex
+		score = score * differentLetterPenaltyWeight
```

Comment:
Agreed. I had been thinking we should detect the same.

Reply:
Great info. Thanks Adam, I'll have a read